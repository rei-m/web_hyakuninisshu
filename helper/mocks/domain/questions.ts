import { Question } from '@src/domain/models';

export const MOCK_QUESTION_1: Question = {
  id: 1,
  correctAnswerKarutaNo: 1,
  choiceKarutaNoList: [1, 2, 3, 4],
};

export const MOCK_QUESTION_2: Question = {
  id: 2,
  correctAnswerKarutaNo: 2,
  choiceKarutaNoList: [1, 2, 3, 4],
};

export const MOCK_QUESTION_1_STARTED: Question = {
  id: 1,
  correctAnswerKarutaNo: 1,
  choiceKarutaNoList: [1, 2, 3, 4],
  startTime: 1583595557624,
};

export const MOCK_QUESTION_1_ANSWERED_CORRECT: Question = {
  id: 1,
  correctAnswerKarutaNo: 1,
  choiceKarutaNoList: [1, 2, 3, 4],
  startTime: 1583595557624,
  answer: {
    isCorrect: true,
    answerTime: 1583595559624,
  },
};

export const MOCK_QUESTION_2_ANSWERED_WRONG: Question = {
  id: 2,
  correctAnswerKarutaNo: 2,
  choiceKarutaNoList: [1, 2, 3, 4],
  startTime: 1583595689123,
  answer: {
    isCorrect: false,
    answerTime: 1583595690123,
  },
};
