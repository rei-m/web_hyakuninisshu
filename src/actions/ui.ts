import { Action, Dispatch } from 'redux';
import { Color, Kimariji } from '@src/types';

export const OPEN_KARUTAS_FILTER_NAME = 'OPEN_KARUTAS_FILTER';
export type OPEN_KARUTAS_FILTER_TYPE = typeof OPEN_KARUTAS_FILTER_NAME;

export const CLOSE_KARUTAS_FILTER_NAME = 'CLOSE_KARUTAS_FILTER';
export type CLOSE_KARUTAS_FILTER_TYPE = typeof CLOSE_KARUTAS_FILTER_NAME;

export const TOGGLE_KARUTAS_COLOR_NAME = 'TOGGLE_KARUTAS_COLOR';
export type TOGGLE_KARUTAS_COLOR_TYPE = typeof TOGGLE_KARUTAS_COLOR_NAME;

export const TOGGLE_KARUTAS_KIMARIJI_NAME = 'TOGGLE_KARUTAS_KIMARIJI_NAME';
export type TOGGLE_KARUTAS_KIMARIJI_TYPE = typeof TOGGLE_KARUTAS_KIMARIJI_NAME;

export interface OpenKarutasFilterAction extends Action {
  readonly type: OPEN_KARUTAS_FILTER_TYPE;
}

export interface CloseKarutasFilterAction extends Action {
  readonly type: CLOSE_KARUTAS_FILTER_TYPE;
}

export interface OKarutasFilterAction extends Action {
  readonly type: OPEN_KARUTAS_FILTER_TYPE;
}

export interface ToggleKarutasColorAction extends Action {
  readonly type: TOGGLE_KARUTAS_COLOR_TYPE;
  readonly payload: {
    readonly color: Color;
    readonly checked: boolean;
  };
}

export interface ToggleKarutasKimarijiAction extends Action {
  readonly type: TOGGLE_KARUTAS_KIMARIJI_TYPE;
  readonly payload: {
    readonly kimariji: Kimariji;
    readonly checked: boolean;
  };
}

export type UiActions =
  | OpenKarutasFilterAction
  | CloseKarutasFilterAction
  | ToggleKarutasColorAction
  | ToggleKarutasKimarijiAction;

/*
 * action creators
 */
export const openKarutasFilter = () => (dispatch: Dispatch<OpenKarutasFilterAction>) =>
  dispatch({
    type: OPEN_KARUTAS_FILTER_NAME,
  });

export const closeKarutasFilter = () => (dispatch: Dispatch<CloseKarutasFilterAction>) =>
  dispatch({
    type: CLOSE_KARUTAS_FILTER_NAME,
  });

export const toggleKarutasColor = (color: Color, checked: boolean) => (
  dispatch: Dispatch<ToggleKarutasColorAction>
) => {
  const action: ToggleKarutasColorAction = {
    payload: {
      color,
      checked,
    },
    type: TOGGLE_KARUTAS_COLOR_NAME,
  };

  dispatch(action);
};

export const toggleKarutasKimariji = (kimariji: Kimariji, checked: boolean) => (
  dispatch: Dispatch<ToggleKarutasKimarijiAction>
) => {
  const action: ToggleKarutasKimarijiAction = {
    payload: {
      kimariji,
      checked,
    },
    type: TOGGLE_KARUTAS_KIMARIJI_NAME,
  };

  dispatch(action);
};
