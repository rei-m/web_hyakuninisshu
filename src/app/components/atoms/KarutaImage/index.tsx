import { karutaNoToJPNText } from '@/domains/models/KarutaNo';

import type { Karuta } from '@/domains/models';

export type KarutaImageProps = {
  karuta: Karuta;
  width: number;
};

export const KarutaImage = ({ karuta, width }: KarutaImageProps) => (
  // eslint-disable-next-line @next/next/no-img-element
  <img src={`/karuta/karuta_${karuta.imageNo}.jpg`} alt={karutaNoToJPNText({ karutaNo: karuta.no })} width={width} />
);
