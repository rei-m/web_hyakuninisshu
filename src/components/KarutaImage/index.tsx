import * as React from 'react';
import { toKarutaIdString } from '@src/components/helper';
// TODO: Spriteする??
import * as karuta001 from './karuta_001.jpg';
import * as karuta002 from './karuta_002.jpg';
import * as karuta003 from './karuta_003.jpg';
import * as karuta004 from './karuta_004.jpg';
import * as karuta005 from './karuta_005.jpg';
import * as karuta006 from './karuta_006.jpg';
import * as karuta007 from './karuta_007.jpg';
import * as karuta008 from './karuta_008.jpg';
import * as karuta009 from './karuta_009.jpg';
import * as karuta010 from './karuta_010.jpg';
import * as karuta011 from './karuta_011.jpg';
import * as karuta012 from './karuta_012.jpg';
import * as karuta013 from './karuta_013.jpg';
import * as karuta014 from './karuta_014.jpg';
import * as karuta015 from './karuta_015.jpg';
import * as karuta016 from './karuta_016.jpg';
import * as karuta017 from './karuta_017.jpg';
import * as karuta018 from './karuta_018.jpg';
import * as karuta019 from './karuta_019.jpg';
import * as karuta020 from './karuta_020.jpg';
import * as karuta021 from './karuta_021.jpg';
import * as karuta022 from './karuta_022.jpg';
import * as karuta023 from './karuta_023.jpg';
import * as karuta024 from './karuta_024.jpg';
import * as karuta025 from './karuta_025.jpg';
import * as karuta026 from './karuta_026.jpg';
import * as karuta027 from './karuta_027.jpg';
import * as karuta028 from './karuta_028.jpg';
import * as karuta029 from './karuta_029.jpg';
import * as karuta030 from './karuta_030.jpg';
import * as karuta031 from './karuta_031.jpg';
import * as karuta032 from './karuta_032.jpg';
import * as karuta033 from './karuta_033.jpg';
import * as karuta034 from './karuta_034.jpg';
import * as karuta035 from './karuta_035.jpg';
import * as karuta036 from './karuta_036.jpg';
import * as karuta037 from './karuta_037.jpg';
import * as karuta038 from './karuta_038.jpg';
import * as karuta039 from './karuta_039.jpg';
import * as karuta040 from './karuta_040.jpg';
import * as karuta041 from './karuta_041.jpg';
import * as karuta042 from './karuta_042.jpg';
import * as karuta043 from './karuta_043.jpg';
import * as karuta044 from './karuta_044.jpg';
import * as karuta045 from './karuta_045.jpg';
import * as karuta046 from './karuta_046.jpg';
import * as karuta047 from './karuta_047.jpg';
import * as karuta048 from './karuta_048.jpg';
import * as karuta049 from './karuta_049.jpg';
import * as karuta050 from './karuta_050.jpg';
import * as karuta051 from './karuta_051.jpg';
import * as karuta052 from './karuta_052.jpg';
import * as karuta053 from './karuta_053.jpg';
import * as karuta054 from './karuta_054.jpg';
import * as karuta055 from './karuta_055.jpg';
import * as karuta056 from './karuta_056.jpg';
import * as karuta057 from './karuta_057.jpg';
import * as karuta058 from './karuta_058.jpg';
import * as karuta059 from './karuta_059.jpg';
import * as karuta060 from './karuta_060.jpg';
import * as karuta061 from './karuta_061.jpg';
import * as karuta062 from './karuta_062.jpg';
import * as karuta063 from './karuta_063.jpg';
import * as karuta064 from './karuta_064.jpg';
import * as karuta065 from './karuta_065.jpg';
import * as karuta066 from './karuta_066.jpg';
import * as karuta067 from './karuta_067.jpg';
import * as karuta068 from './karuta_068.jpg';
import * as karuta069 from './karuta_069.jpg';
import * as karuta070 from './karuta_070.jpg';
import * as karuta071 from './karuta_071.jpg';
import * as karuta072 from './karuta_072.jpg';
import * as karuta073 from './karuta_073.jpg';
import * as karuta074 from './karuta_074.jpg';
import * as karuta075 from './karuta_075.jpg';
import * as karuta076 from './karuta_076.jpg';
import * as karuta077 from './karuta_077.jpg';
import * as karuta078 from './karuta_078.jpg';
import * as karuta079 from './karuta_079.jpg';
import * as karuta080 from './karuta_080.jpg';
import * as karuta081 from './karuta_081.jpg';
import * as karuta082 from './karuta_082.jpg';
import * as karuta083 from './karuta_083.jpg';
import * as karuta084 from './karuta_084.jpg';
import * as karuta085 from './karuta_085.jpg';
import * as karuta086 from './karuta_086.jpg';
import * as karuta087 from './karuta_087.jpg';
import * as karuta088 from './karuta_088.jpg';
import * as karuta089 from './karuta_089.jpg';
import * as karuta090 from './karuta_090.jpg';
import * as karuta091 from './karuta_091.jpg';
import * as karuta092 from './karuta_092.jpg';
import * as karuta093 from './karuta_093.jpg';
import * as karuta094 from './karuta_094.jpg';
import * as karuta095 from './karuta_095.jpg';
import * as karuta096 from './karuta_096.jpg';
import * as karuta097 from './karuta_097.jpg';
import * as karuta098 from './karuta_098.jpg';
import * as karuta099 from './karuta_099.jpg';
import * as karuta100 from './karuta_100.jpg';

const KARUTA_IMAGE_LIST = [
  karuta001,
  karuta002,
  karuta003,
  karuta004,
  karuta005,
  karuta006,
  karuta007,
  karuta008,
  karuta009,
  karuta010,
  karuta011,
  karuta012,
  karuta013,
  karuta014,
  karuta015,
  karuta016,
  karuta017,
  karuta018,
  karuta019,
  karuta020,
  karuta021,
  karuta022,
  karuta023,
  karuta024,
  karuta025,
  karuta026,
  karuta027,
  karuta028,
  karuta029,
  karuta030,
  karuta031,
  karuta032,
  karuta033,
  karuta034,
  karuta035,
  karuta036,
  karuta037,
  karuta038,
  karuta039,
  karuta040,
  karuta041,
  karuta042,
  karuta043,
  karuta044,
  karuta045,
  karuta046,
  karuta047,
  karuta048,
  karuta049,
  karuta050,
  karuta051,
  karuta052,
  karuta053,
  karuta054,
  karuta055,
  karuta056,
  karuta057,
  karuta058,
  karuta059,
  karuta060,
  karuta061,
  karuta062,
  karuta063,
  karuta064,
  karuta065,
  karuta066,
  karuta067,
  karuta068,
  karuta069,
  karuta070,
  karuta071,
  karuta072,
  karuta073,
  karuta074,
  karuta075,
  karuta076,
  karuta077,
  karuta078,
  karuta079,
  karuta080,
  karuta081,
  karuta082,
  karuta083,
  karuta084,
  karuta085,
  karuta086,
  karuta087,
  karuta088,
  karuta089,
  karuta090,
  karuta091,
  karuta092,
  karuta093,
  karuta094,
  karuta095,
  karuta096,
  karuta097,
  karuta098,
  karuta099,
  karuta100
];

export interface KarutaImageProps {
  readonly karutaId: number;
  readonly style?: React.CSSProperties;
}

const KarutaImage: React.SFC<KarutaImageProps> = ({ karutaId, style }) => (
  <img
    src={KARUTA_IMAGE_LIST[karutaId - 1]}
    alt={toKarutaIdString(karutaId)}
    style={style}
  />
);

export default KarutaImage;
