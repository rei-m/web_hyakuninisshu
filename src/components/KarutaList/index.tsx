import * as React from 'react';
import styled from 'styled-components';
import { withAppTheme } from '../../styles';
import { Karuta } from '../../types';
import KarutaListRow from './KarutaListRow';

export interface KarutaListProps {
  readonly karutas: Karuta[];
  readonly onClickRow: (karutaId: number) => void;
}

const Root = withAppTheme(styled.div)`
  display: flex;
  flex-wrap: wrap;
  margin: auto;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  @media screen and (min-width: ${({ theme }) => theme.minWidthWide}) {
    flex-direction: row;
  }
`;

const KarutaList: React.SFC<KarutaListProps> = ({ karutas, onClickRow }) => (
  <Root>
    {karutas.map((k, i) => (
      <KarutaListRow
        karuta={k}
        key={i}
        onClickRow={onClickRow}
        data-test={`row-${k.id}`}
      />
    ))}
  </Root>
);

export default KarutaList;
