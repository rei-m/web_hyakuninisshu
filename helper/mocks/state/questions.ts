import { questionsTypes } from '@src/state/questions';

export const MOCK_YOMIFUDA_1: questionsTypes.YomiFuda = {
  karutaNo: 1,
  shoku: '秋の田の',
  niku: 'かりほの庵の',
  sanku: '苫をあらみ',
};

export const MOCK_TORIFUDA_1: questionsTypes.ToriFuda = {
  karutaNo: 1,
  shiku: 'きりにぬれつつ',
  kekku: 'わがころもでは',
};

export const MOCK_TORIFUDA_2: questionsTypes.ToriFuda = {
  karutaNo: 2,
  shiku: 'ころもほすてふ',
  kekku: 'あまのかぐやま',
};

export const MOCK_TORIFUDA_3: questionsTypes.ToriFuda = {
  karutaNo: 3,
  shiku: 'ながながしよを',
  kekku: 'ひとりかもねむ',
};

export const MOCK_TORIFUDA_4: questionsTypes.ToriFuda = {
  karutaNo: 4,
  shiku: 'ふじのたかねに',
  kekku: 'ゆきはふりつつ',
};
