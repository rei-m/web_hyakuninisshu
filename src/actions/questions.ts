import { Action, Dispatch } from 'redux';
import { GlobalState } from '@src/state';
import { Answer, Karuta, Question } from '@src/types';
import { randomizeArray } from '@src/utils';
import { fetchTorifudas, questionsFilter, toDulation } from '@src/utils/questions';
import {
  ColorCondition,
  KarutaStyleCondition,
  KimarijiCondition,
  QuestionAnimCondition,
  QuestionState,
  RangeFromCondition,
  RangeToCondition,
} from '@src/enums';

export const START_TRAINING_NAME = 'START_TRAINING_NAME';
export type START_TRAINING_TYPE = typeof START_TRAINING_NAME;

export const START_EXAM_NAME = 'START_EXAM_NAME';
export type START_EXAM_TYPE = typeof START_EXAM_NAME;

export const RESTART_QUESTIONS_NAME = 'RESTART_QUESTIONS_NAME';
export type RESTART_QUESTIONS_TYPE = typeof RESTART_QUESTIONS_NAME;

export const ANSWER_QUESTION_NAME = 'ANSWER_QUESTION_NAME';
export type ANSWER_QUESTION_TYPE = typeof ANSWER_QUESTION_NAME;

export const CONFIRM_CORRECT_NAME = 'CONFIRM_CORRECT_NAME';
export type CONFIRM_CORRECT_TYPE = typeof CONFIRM_CORRECT_NAME;

export const OPEN_NEXT_QUESTION_NAME = 'OPEN_NEXT_QUESTION_NAME';
export type OPEN_NEXT_QUESTION_TYPE = typeof OPEN_NEXT_QUESTION_NAME;

export const FINISH_QUESTIONS_NAME = 'FINISH_QUESTIONS_NAME';
export type FINISH_QUESTIONS_TYPE = typeof FINISH_QUESTIONS_NAME;

export interface StartTrainingAction extends Action {
  readonly type: START_TRAINING_TYPE;
  readonly payload: {
    readonly karutas: Karuta[];
    readonly questions: Question[];
    readonly startedTime: number;
    readonly nextState: QuestionState;
    readonly dulation: number;
  };
  readonly meta: {
    readonly rangeFrom: RangeFromCondition;
    readonly rangeTo: RangeToCondition;
    readonly kimariji: KimarijiCondition;
    readonly color: ColorCondition;
    readonly kamiNoKuStyle: KarutaStyleCondition;
    readonly shimoNoKuStyle: KarutaStyleCondition;
    readonly questionAnim: QuestionAnimCondition;
  };
}

export interface StartExamAction extends Action {
  readonly type: START_EXAM_TYPE;
  readonly payload: {
    readonly karutas: Karuta[];
    readonly questions: Question[];
    readonly startedTime: number;
    readonly nextState: QuestionState;
  };
}

export interface RestartQuestionsAction extends Action {
  readonly type: RESTART_QUESTIONS_TYPE;
  readonly payload: {
    readonly questions: Question[];
    readonly startedTime: number;
    readonly nextState: QuestionState;
  };
}

export interface AnswerQuestionAction extends Action {
  readonly type: ANSWER_QUESTION_TYPE;
  readonly payload: {
    readonly answer: Answer;
    readonly nextState: QuestionState;
  };
}

export interface ConfirmCorrectAction extends Action {
  readonly type: CONFIRM_CORRECT_TYPE;
  readonly payload: {
    readonly nextState: QuestionState;
  };
}

export interface OpenNextQuestionAction extends Action {
  readonly type: OPEN_NEXT_QUESTION_TYPE;
  readonly payload: {
    readonly nextIndex: number;
    readonly nextState: QuestionState;
    readonly startedTime: number;
  };
}

export interface FinishQuestionsAction extends Action {
  readonly type: FINISH_QUESTIONS_TYPE;
  readonly payload: {
    readonly nextState: QuestionState;
  };
}

export type QuestionsActions =
  | StartTrainingAction
  | StartExamAction
  | RestartQuestionsAction
  | AnswerQuestionAction
  | ConfirmCorrectAction
  | OpenNextQuestionAction
  | FinishQuestionsAction;

/*
 * action creators
 */
export const startTraining = (
  karutas: Karuta[],
  rangeFrom: RangeFromCondition,
  rangeTo: RangeToCondition,
  kimariji: KimarijiCondition,
  color: ColorCondition,
  kamiNoKuStyle: KarutaStyleCondition,
  shimoNoKuStyle: KarutaStyleCondition,
  questionAnim: QuestionAnimCondition
) => (dispatch: Dispatch<QuestionsActions>) => {
  const questions = new QuestionsFactory(karutas)
    .setRange(rangeFrom, rangeTo)
    .setKimariji(kimariji)
    .setColor(color)
    .setKamiNoKuStyle(kamiNoKuStyle)
    .setShimoNoKuStyle(shimoNoKuStyle)
    .create();

  const action: StartTrainingAction = {
    meta: {
      color,
      kamiNoKuStyle,
      kimariji,
      rangeFrom,
      rangeTo,
      shimoNoKuStyle,
      questionAnim,
    },
    payload: {
      karutas,
      nextState: QuestionState.InAnswer,
      questions: randomizeArray<Question>(questions),
      startedTime: new Date().getTime(),
      dulation: toDulation(questionAnim),
    },
    type: START_TRAINING_NAME,
  };

  dispatch(action);
};

export const startExam = (karutas: Karuta[]) => (dispatch: Dispatch<QuestionsActions>) => {
  const questions = new QuestionsFactory(karutas)
    .setRange(RangeFromCondition.One, RangeToCondition.OneHundred)
    .setKimariji(KimarijiCondition.None)
    .setColor(ColorCondition.None)
    .setKamiNoKuStyle(KarutaStyleCondition.KanjiAndKana)
    .setShimoNoKuStyle(KarutaStyleCondition.KanaOnly)
    .create();

  const action: StartExamAction = {
    payload: {
      karutas,
      nextState: QuestionState.InAnswer,
      questions: randomizeArray(questions),
      startedTime: new Date().getTime(),
    },
    type: START_EXAM_NAME,
  };

  dispatch(action);
};

export const restartQuestions = () => (dispatch: Dispatch<QuestionsActions>, getState: () => GlobalState) => {
  const { questions, answers } = getState().questions;
  // TODO: 全て回答済みでなかったらエラー
  const finder: { [questionId: number]: Question } = questions.reduce((previous, current) => {
    return { ...previous, [current.id]: current };
  }, {});
  const targets = answers
    .filter(a => !a.correct)
    .map(a => finder[a.questionId])
    .map(q => ({ ...q, toriFudas: randomizeArray(q.toriFudas) }));

  const action: RestartQuestionsAction = {
    payload: {
      nextState: QuestionState.InAnswer,
      questions: randomizeArray(targets),
      startedTime: new Date().getTime(),
    },
    type: RESTART_QUESTIONS_NAME,
  };

  dispatch(action);
};

export const answerQuestion = (questionId: number, karutaNo: number) => (
  dispatch: Dispatch<QuestionsActions>,
  getState: () => GlobalState
) => {
  const { questions, lastStartedTime } = getState().questions;
  const question = questions.find(q => q.id === questionId)!;
  // TODO: QuestionとlastStartedTimeがundefinedの場合
  const correct = question.correctKaruta.no === karutaNo;
  const time = new Date().getTime() - lastStartedTime!;
  const answer = {
    correct,
    karutaNo,
    questionId,
    time,
  };
  const action: AnswerQuestionAction = {
    payload: {
      answer,
      nextState: QuestionState.Answered,
    },
    type: ANSWER_QUESTION_NAME,
  };

  dispatch(action);
};

export const confirmCorrect = (): ConfirmCorrectAction => {
  return {
    payload: {
      nextState: QuestionState.ConfirmCorrect,
    },
    type: CONFIRM_CORRECT_NAME,
  };
};

export const openNextQuestion = () => (dispatch: Dispatch<QuestionsActions>, getState: () => GlobalState) => {
  const { currentIndex } = getState().questions;
  const nextIndex = currentIndex + 1;
  const action: OpenNextQuestionAction = {
    payload: {
      nextIndex,
      nextState: QuestionState.InAnswer,
      startedTime: new Date().getTime(),
    },
    type: OPEN_NEXT_QUESTION_NAME,
  };
  dispatch(action);
};

export const finishQuestions = (): FinishQuestionsAction => {
  return {
    payload: {
      nextState: QuestionState.Finished,
    },
    type: FINISH_QUESTIONS_NAME,
  };
};

class QuestionsFactory {
  private karutas: Karuta[];
  private rangeFrom: number;
  private rangeTo: number;
  private kimariji: number;
  private color: string;
  private kamiNoKuStyle: number;
  private shimoNoKuStyle: number;

  constructor(karutas: Karuta[]) {
    this.karutas = karutas;
    this.rangeFrom = RangeFromCondition.One;
    this.rangeTo = RangeToCondition.OneHundred;
    this.kimariji = KimarijiCondition.None;
    this.color = ColorCondition.None;
    this.kamiNoKuStyle = KarutaStyleCondition.KanjiAndKana;
    this.shimoNoKuStyle = KarutaStyleCondition.KanaOnly;
  }

  public setRange(rangeFrom: number, rangeTo: number) {
    this.rangeFrom = rangeFrom;
    this.rangeTo = rangeTo;
    return this;
  }

  public setKimariji(kimariji: number) {
    this.kimariji = kimariji;
    return this;
  }

  public setColor(color: string) {
    this.color = color;
    return this;
  }

  public setKamiNoKuStyle(kamiNoKuStyle: number) {
    this.kamiNoKuStyle = kamiNoKuStyle;
    return this;
  }

  public setShimoNoKuStyle(shimoNoKuStyle: number) {
    this.shimoNoKuStyle = shimoNoKuStyle;
    return this;
  }

  public create() {
    const targetKarutas = questionsFilter(this.karutas)(this.rangeFrom, this.rangeTo)(this.kimariji)(this.color);

    return targetKarutas.map((k, i) => {
      const id = i + 1;
      const correctKaruta = k;
      const yomiFuda =
        this.kamiNoKuStyle === 0
          ? {
              firstText: k.firstKanji,
              karutaNo: k.no,
              questionId: id,
              secondText: k.secondKanji,
              thirdText: k.thirdKanji,
            }
          : {
              firstText: k.firstKana,
              karutaNo: k.no,
              questionId: id,
              secondText: k.secondKana,
              thirdText: k.thirdKana,
            };
      const toriFudas = fetchTorifudas(this.karutas, correctKaruta).map(toriFuda => {
        return this.shimoNoKuStyle === 0
          ? {
              fifthText: toriFuda.fifthKanji,
              fourthText: toriFuda.fourthKanji,
              karutaNo: toriFuda.no,
              questionId: id,
            }
          : {
              fifthText: toriFuda.fifthKana,
              fourthText: toriFuda.fourthKana,
              karutaNo: toriFuda.no,
              questionId: id,
            };
      });
      return { id, correctKaruta, yomiFuda, toriFudas };
    });
  }
}
