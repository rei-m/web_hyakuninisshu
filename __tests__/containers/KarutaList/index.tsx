import * as ReactTestRenderer from 'react-test-renderer';
import * as React from 'react';
import { shallow } from 'enzyme';
import { create } from '@test/factories';
import { sel, setUpQuery } from '@test/helpers';
import { mapStateToProps, KarutaList, Props } from '@src/containers/KarutaList';
import { QueryData } from '@src/components/KarutaImage';
import { initialState as questionsState } from '@src/state/questions';
import { initialState as uiState } from '@src/state/ui';
import { Color, Karuta, Kimariji } from '@src/types';
import { GlobalState } from '@src/state';

describe('<KarutaList />', () => {
  describe('components', () => {
    let props: Props;

    beforeEach(() => {
      setUpQuery<QueryData>({
        karutaImages: {
          edges: [
            {
              node: {
                childImageSharp: {
                  fluid: {
                    aspectRatio: 1,
                    sizes: `100 200 300`,
                    src: `karuta_001.jpg`,
                    srcSet: `asdfasdf`,
                  },
                },
              },
            },
            {
              node: {
                childImageSharp: {
                  fluid: {
                    aspectRatio: 1,
                    sizes: `100 200 300`,
                    src: `karuta_002.jpg`,
                    srcSet: `asdfasdf`,
                  },
                },
              },
            },
          ],
        },
      });

      props = {
        karutas: [
          create<Karuta>('karuta', {
            id: '1',
            no: 1,
          }),
          create<Karuta>('karuta', {
            id: '2',
            no: 2,
          }),
        ],
        onClickRow: jest.fn(),
      };
    });

    it('should render component', () => {
      const renderer = ReactTestRenderer.create(<KarutaList {...props} />);
      expect(renderer.toJSON()).toMatchSnapshot();
    });

    it('should fire onClickRow when karuta clicked', () => {
      const wrapper = shallow(<KarutaList {...props} />)
        .find(sel('row-1'))
        .dive()
        .dive();
      wrapper.simulate('click');
      expect(props.onClickRow).toHaveBeenCalledWith(1);
    });
  });

  describe('mapStateToProps', () => {
    it('should convert state to props', () => {
      const karutas: Karuta[] = Array.from(Array(100).keys()).map(i =>
        create<Karuta>('karuta', {
          color: (i < 20 ? 'blue' : 'pink') as Color,
          id: (i + 1).toString(),
          no: i + 1,
          kimariji: ((i % 5) + 1) as Kimariji,
        })
      );

      const props: Props = {
        karutas,
        onClickRow: jest.fn(),
      };

      const state: GlobalState = {
        questions: questionsState,
        ui: {
          ...uiState,
          karutasFilter: {
            colors: [{ color: 'blue', checked: true }, { color: 'pink', checked: false }],
            kimarijis: [{ kimariji: 1, checked: true }, { kimariji: 2, checked: false }],
            open: false,
          },
        },
      };

      expect(mapStateToProps(state, props)).toMatchSnapshot();
    });
  });
});
