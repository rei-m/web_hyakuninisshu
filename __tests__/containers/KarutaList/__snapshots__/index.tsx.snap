// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KarutaList /> components should render component 1`] = `
<div
  className="sc-htoDjs eacTSX"
>
  <div
    className="sc-bdVaJa dzVtvu"
    onClick={[Function]}
  >
    <div
      className="sc-bwzfXH fnStwP"
    >
      <div
        className="sc-bxivhb ivoHNN"
      >
        一番
      </div>
      <div
        className=" gatsby-image-wrapper"
        style={
          Object {
            "margin": "auto",
            "overflow": "hidden",
            "position": "relative",
            "width": 40,
          }
        }
      >
        <div
          style={
            Object {
              "paddingBottom": "100%",
              "width": "100%",
            }
          }
        />
        <picture>
          <source
            sizes="100 200 300"
            srcSet="asdfasdf"
          />
          <img
            alt="一番"
            onLoad={[Function]}
            src="karuta_001.jpg"
            style={
              Object {
                "height": "100%",
                "left": 0,
                "objectFit": "cover",
                "objectPosition": "center",
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "transition": "opacity 0.5s",
                "width": "100%",
              }
            }
          />
        </picture>
        <noscript
          dangerouslySetInnerHTML={
            Object {
              "__html": "<picture><source srcSet=\\"asdfasdf\\" sizes=\\"100 200 300\\" /><img src=\\"karuta_001.jpg\\" alt=\\"一番\\" style=\\"position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
            }
          }
        />
      </div>
    </div>
    <div
      className="sc-htpNat Cyytt"
    >
      <div
        className="sc-ifAKCX TiiUY"
      >
        初句 二句 三句
      </div>
      <div
        className="sc-ifAKCX TiiUY"
      >
        四句 五句
      </div>
      <div
        className="sc-EHOje kQvhir"
      >
        詠み手
      </div>
    </div>
  </div>
  <div
    className="sc-bdVaJa dzVtvu"
    onClick={[Function]}
  >
    <div
      className="sc-bwzfXH fnStwP"
    >
      <div
        className="sc-bxivhb ivoHNN"
      >
        二番
      </div>
      <div
        className=" gatsby-image-wrapper"
        style={
          Object {
            "margin": "auto",
            "overflow": "hidden",
            "position": "relative",
            "width": 40,
          }
        }
      >
        <div
          style={
            Object {
              "paddingBottom": "100%",
              "width": "100%",
            }
          }
        />
        <picture>
          <source
            sizes="100 200 300"
            srcSet="asdfasdf"
          />
          <img
            alt="二番"
            onLoad={[Function]}
            src="karuta_002.jpg"
            style={
              Object {
                "height": "100%",
                "left": 0,
                "objectFit": "cover",
                "objectPosition": "center",
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "transition": "opacity 0.5s",
                "width": "100%",
              }
            }
          />
        </picture>
        <noscript
          dangerouslySetInnerHTML={
            Object {
              "__html": "<picture><source srcSet=\\"asdfasdf\\" sizes=\\"100 200 300\\" /><img src=\\"karuta_002.jpg\\" alt=\\"二番\\" style=\\"position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
            }
          }
        />
      </div>
    </div>
    <div
      className="sc-htpNat Cyytt"
    >
      <div
        className="sc-ifAKCX TiiUY"
      >
        初句 二句 三句
      </div>
      <div
        className="sc-ifAKCX TiiUY"
      >
        四句 五句
      </div>
      <div
        className="sc-EHOje kQvhir"
      >
        詠み手
      </div>
    </div>
  </div>
</div>
`;

exports[`<KarutaList /> mapStateToProps should convert state to props 1`] = `
Object {
  "karutas": Array [
    Object {
      "color": "blue",
      "creator": "詠み手",
      "fifthKana": "ごく",
      "fifthKanji": "五句",
      "firstKana": "しょく",
      "firstKanji": "初句",
      "fourthKana": "よんく",
      "fourthKanji": "四句",
      "id": "1",
      "imageNo": "001",
      "kimariji": 1,
      "no": 1,
      "secondKana": "にく",
      "secondKanji": "二句",
      "thirdKana": "さんく",
      "thirdKanji": "三句",
      "translation": "歌の訳",
    },
    Object {
      "color": "blue",
      "creator": "詠み手",
      "fifthKana": "ごく",
      "fifthKanji": "五句",
      "firstKana": "しょく",
      "firstKanji": "初句",
      "fourthKana": "よんく",
      "fourthKanji": "四句",
      "id": "6",
      "imageNo": "001",
      "kimariji": 1,
      "no": 6,
      "secondKana": "にく",
      "secondKanji": "二句",
      "thirdKana": "さんく",
      "thirdKanji": "三句",
      "translation": "歌の訳",
    },
    Object {
      "color": "blue",
      "creator": "詠み手",
      "fifthKana": "ごく",
      "fifthKanji": "五句",
      "firstKana": "しょく",
      "firstKanji": "初句",
      "fourthKana": "よんく",
      "fourthKanji": "四句",
      "id": "11",
      "imageNo": "001",
      "kimariji": 1,
      "no": 11,
      "secondKana": "にく",
      "secondKanji": "二句",
      "thirdKana": "さんく",
      "thirdKanji": "三句",
      "translation": "歌の訳",
    },
    Object {
      "color": "blue",
      "creator": "詠み手",
      "fifthKana": "ごく",
      "fifthKanji": "五句",
      "firstKana": "しょく",
      "firstKanji": "初句",
      "fourthKana": "よんく",
      "fourthKanji": "四句",
      "id": "16",
      "imageNo": "001",
      "kimariji": 1,
      "no": 16,
      "secondKana": "にく",
      "secondKanji": "二句",
      "thirdKana": "さんく",
      "thirdKanji": "三句",
      "translation": "歌の訳",
    },
  ],
  "onClickRow": [MockFunction],
}
`;
